### PR TITLE
Keep track of remote object reference count.

### LINKED ISSUE #9387: Crash in electron 1.6.8 using remote when sandboxed
Crash in electron remote.
* Electron version: 1.6.8
* Operating system: Mac OSX 10.10.5

This is only happening on 1.6.8.  Doesn't happen in 1.6.7.

Repo:
1. git clone https://github.com/lneir/electron-remote-issue
2. follow instructions in README
3. crash stack shown below.

This demo repo has sandboxing turned on and is using preload script to access remote class.

crash stack from renderer console:
Uncaught Error: Cannot call constructor on missing remote object 36
Error: Cannot call constructor on missing remote object 36
    at throwRPCError (/Users/lneir/IntelliJProjects/tests/electron-remote-issue/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Users/lneir/IntelliJProjects/tests/electron-remote-issue/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:303:7)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at WebContents.<anonymous> (/Users/lneir/IntelliJProjects/tests/electron-remote-issue/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:256:37)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at throwRPCError (/Users/lneir/IntelliJProjects/tests/electron-remote-issue/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:143:17)
    at EventEmitter.<anonymous> (/Users/lneir/IntelliJProjects/tests/electron-remote-issue/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/rpc-server.js:303:7)
    at emitThree (events.js:116:13)
    at EventEmitter.emit (events.js:194:7)
    at WebContents.<anonymous> (/Users/lneir/IntelliJProjects/tests/electron-remote-issue/node_modules/electron/dist/Electron.app/Contents/Resources/electron.asar/browser/api/web-contents.js:256:37)
    at emitTwo (events.js:106:13)
    at WebContents.emit (events.js:191:7)
    at metaToValue (<anonymous>:552:13)
    at new remoteFunction (<anonymous>:565:20)
    at HTMLButtonElement.<anonymous> (file:///Users/lneir/IntelliJProjects/tests/electron-remote-issue/index.html:15:17)

### PR DESCRIPTION
**Pull request by tarruda**
This is required to ensure page reloads are handled properly.

Close #9387

### CODE REVIEW

File: lib/browser/objects-registry.js [hunk 1]
```diff
@@ -28,11 +28,9 @@ class ObjectsRegistry {
       owner = this.owners[webContentsId] = new Set()
       this.registerDeleteListener(webContents, webContentsId)
     }
-    if (!owner.has(id)) {
```
Comment (zcbenz) [inline lib/browser/objects-registry.js]: Both `remote` and `ObjectsRegistry` assume that one `webContents` can only hold one reference to an object in the main process, so when `remote` requests the same object for multiple times in one renderer process, the reference count to the object only adds one. The `if (!owner.has(id))` condition is to guard this assumption.

If we change this, the object would be leaked in the main process when a `webContents` requested the same remote object for multiple times.

Regarding the issue #9387:

```
initial load
hello.js was added to objects-registry and referenced count increased
reload
hello.js was added to objects-registry(no reference count increased)
```

The root cause is: after page was reloaded all remote objects referenced by the reloaded `webContents` should be cleared, but for some reason it is not true for sandboxed `webContents`.

The clearance is done by listening to the `render-view-deleted` event of `webContents` in the `registerDeleteListener` method, the bug is probably caused by the `render-view-deleted` event not emitting when reloading sandboxed `webContents`.

Comment (tarruda) [inline lib/browser/objects-registry.js]: Thanks for the explanation, I will fix the PR to ensure render-view-deleted is emitted/handled properly in sandbox

**Comment by tarruda** [comment]
@zcbenz it turns out `render-view-deleted` was never called because sandbox reuses the same process for reload, so the fix was to use the same `ChildWebContentsTracker` trick used by `nativeWindowOpen` to force a process respawn on reload of the top-level window.

This creates another question though: Won't this be a problem for child windows given that they also have dedicated `webContents` that can be reloaded without creating new processes? If so, maybe we should fix the assumption of one webContents only having one reference in objects-registry.js?

**Comment by zcbenz** [comment]
@tarruda We can probably solve this by sending an sync message to the main process when a web page's V8 context is going to be released, and clear the `ObjectsRegistry` when received that event instead of listening to `render-view-deleted`.

> maybe we should fix the assumption of one webContents only having one reference in objects-registry.js?

This would result in the same object in main process having different remote objects, e.g. in renderer process `remote.require('module') !== remote.require('module')`.

**Comment by tarruda** [comment]
@zcbenz I have fixed according to your specifications, but sending a synchronous message before navigation seems risky due to the possibility of deadlocks(I remember reading somewhere in chrome docs that it is a good practice to avoid relying on synchronous messages, if possible). Also it doesn't seem very robust since if the renderer crashes no message will be sent(assuming "render-view-deleted" is emitted even if the process crashes).

IMHO I think it is cleaner if we simply assume one webcontents can reference the same object multiple times, I don't see how `remote.require('module') !== remote.require('module')` could happen in this case. If the renderer receives the same id for an object, it would simply fetch the existing object from its `remoteObjectCache`, no?

As for a solution, I imagine the following: We maintain a global per-object reference count(all references to the object across all webcontents) plus a webcontents-specific per-object reference count. If a webcontents is ever killed of removed before all objects referenced by it are explicitly garbage collected, we simply subtract the webcontents count from the global count. If you are ok with it, I can give a shot at implementing this.

**Comment by zcbenz** [comment]
> but sending a synchronous message before navigation seems risky due to the possibility of deadlocks

It depends on the direction of the message, sending sync message from renderer process to main process is totally fine, and the `remote` module has been doing it very heavily.

> IMHO I think it is cleaner if we simply assume one webcontents can reference the same object multiple times, I don't see how remote.require('module') !== remote.require('module') could happen in this case. If the renderer receives the same id for an object, it would simply fetch the existing object from its remoteObjectCache, no?

The reference count of remote object in renderer process is only subtracted when it gets garbage collected, currently it works like this:

```js
let a = remote.require('module')
// referenceCount +=1 
let b = remote.require('module')
// the object exists in remoteCache, so referenceCount is not changed
a === b

a = null
gc()
// the object is still referenced by b, so referenceCount is not changed

b = null
gc()
// referenceCount -=1

// referenceCount ==0, object is released.
```

If we add reference count whenever an object is requested, the there will have leak:

```js
let a = remote.require('module')
// referenceCount +=1 
let b = remote.require('module')
// referenceCount +=1 
a === b

a = null
gc()
// the object is still referenced by b, so referenceCount is not changed

b = null
gc()
// garbage collection happened, referenceCount -=1

// referenceCount == 1, the object is leaked.
```

Unless we don't cache remote objects and create a proxy whenever an object is requested:

```js
let a = remote.require('module')
// referenceCount +=1 
let b = remote.require('module')
// referenceCount +=1 
a !== b

a = null
gc()
// garbage collection happened, referenceCount -=1

b = null
gc()
// garbage collection happened, referenceCount -=1

// referenceCount ==0, object is released.
```

But it breaks object equality and consumes much more resources, (operations like `remote.getCurrentWindow()` would create a huge proxy object every time).

**Comment by tarruda** [comment]
OK @zcbenz I understand the issue now, thanks again for explaining.

I have added the synchronous message to release references but didn't remove the `render-view-deleted` listener since that will be fired even if the renderer is terminated abnormally. It should be fine since the `clear` method guards against double release.

**Comment by lneir** [comment]
any updates when this will be merged?

**Pull request merged**